### DISCUSSION TITLE
Trivy throwing false positives when NPM overrides are used

### DISCUSSION THREAD

**Discussion by vhscom** [aquasecurity/trivy#3974] [Q&A]
## Description

This is a continuation of #2831 scoped down to only NPM. Given the way Trivy runs even a [standardized npm override](https://github.com/npm/rfcs/blob/main/accepted/0036-overrides.md) will not satisfy the scanner.

## What did you expect to happen?
If a NPM override is made in a `package.json` file, Trivy picks up on that and adapts its output accordingly.

## What happened instead?
Trivy reports vulnerabilities even after they've been addressed using standardized means.

There well may be a reason Trivy cannot support or choses not to support NPM overrides. But I haven't heard from maintainers as to whether or not support for NPM overrides is planned or if it's considered a `#wontfix`. That amount of info should be enough for most to decide if Trivy is right for their stacks using NPM and help them understand how to handle false positives.

**Comment by DmitriyLewen** [comment]
Hello @vhscom 
Thanks for your report!

You are right. Trivy doesn't currently support parsing overridden dependencies.
It used to be difficult to create such changes, but now we have PostAnalyzers to deal with such cases.

I will create PR to add this logic. When PR is merged, I will write to you.

Regards, Dmitriy

**Comment by DmitriyLewen** [comment]
Hello @vhscom 
Can you help me understand:
I played with [overrides](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides) field in `package.json` file.
1) Trivy only parses project information for `package.json` file, then `overrides` doesn't affect on this file.
2) `package-lock.json` file uses overwritten dependency version, then Trivy should detect correctly version.
e.g.:
`package.json` file:
```
➜  cat package.json
{
...
  "dependencies": {
    "debug": "^2.6.9"
  }
}
➜  cat package-lock.json| grep ms -A 1
...
    "node_modules/ms": {
      "version": "2.0.0",
...
```
i overwrite `ms` version:
```
➜  3816 cat package.json              
{
...
  "dependencies": {
    "debug": "^2.6.9"
  },
  "overrides": {
    "ms": "1.0.0"
  }
}
➜  npm i --package-lock-only
...
➜  cat package-lock.json| grep ms -A 1
...
    "node_modules/debug/node_modules/ms": {
      "version": "1.0.0",
...
```

**Comment by knqyf263** [comment]
I suspect that the package-lock.json was not re-generated after adding overrides.

**Comment by DmitriyLewen** [comment]
> I suspect that the package-lock.json was not re-generated after adding overrides.

Then `package-lock.json` doesn't contain up-to-date and correct dependencies.

**Comment by knqyf263** [comment]
>Then package-lock.json doesn't contain up-to-date and correct dependencies.

Yes, that is my suspicion. Let's wait for @vhscom's answer.

**Comment by aggieben** [comment]
I can't post code, but I definitely have a current example of trivy relying on versions in the `package.json` and not catching resolved override versions in the `package-lock.json`.  The lockfile is up-to-date.

**Comment by DmitriyLewen** [comment]
Hello @aggieben 
Thanks for your comment!

Can you check your `package-lock.json files`: It should use overridden version(as in my example - https://github.com/aquasecurity/trivy/issues/3816#issuecomment-1475633365).

**Reply by aggieben** [reply]
My recollection at the time was that the `package-lock.json` did, in fact, reflect the overrides correctly.
